### PR TITLE
Building and deploying Third Person Shooter

### PR DESCRIPTION
**Pull request by padhansell**
**Description**
This PR adds nightly build/deploy capability to the project. The pipeline here https://buildkite.com/improbable/unrealgdkthirdpersonshooter-nightly has been set up to build daily at 7:00am against @joshuahuburn's unified-project branch. Once this branch has been merged to master, the pipeline will need to be updated to build from that.

The pipeline will build the GDK for Windows and Linux and will deploy the project to Spatial for QA testing

### CODE REVIEW

**Comment by tenevdev** [comment]
The current build queue will need to be updated as it points to a canary I used for debugging the toolbelt issue. See https://github.com/improbable/platform/pull/8738

**Comment by padhansell** [comment]
@m-samiec @claireimprobable This is now ready for review

**Comment by padhansell** [comment]
Note that the premerge pipeline is still showing in the checks above but has actually been deleted (in favour of a nightly build)

File: ci/setup-and-build.ps1 [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        popd
+    Finish-Event "set-up-gdk-plugin" "build-gdk-third-person-shooter-:windows:"
+
+    # Set LINUX_MULTIARCH_ROOT and then reload it for this script
+    $clang_path = "$($game_home)\UnrealEngine\ClangToolchain\"
+    [Environment]::SetEnvironmentVariable("LINUX_MULTIARCH_ROOT", $clang_path, "Machine")
+    $env:LINUX_MULTIARCH_ROOT = [System.Environment]::GetEnvironmentVariable("LINUX_MULTIARCH_ROOT", "Machine")
+
+    # Allow the GDK plugin to find the engine
+    $env:UNREAL_HOME = "$($game_home)\UnrealEngine\"
```
Comment (m-samiec) [inline ci/setup-and-build.ps1]: This is already set above.

File: ci/setup-and-build.ps1 [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            $gdk_home = Convert-Path .
+            $msbuild_exe = "${env:ProgramFiles(x86)}\MSBuild\14.0\bin\MSBuild.exe"
+
+            # Invoke the setup script
+            &"$($game_home)\Game\Plugins\UnrealGDK\ci\setup-gdk.ps1"
+        popd
+    Finish-Event "set-up-gdk-plugin" "build-gdk-third-person-shooter-:windows:"
+
+    # Set LINUX_MULTIARCH_ROOT and then reload it for this script
+    $clang_path = "$($game_home)\UnrealEngine\ClangToolchain\"
```
Comment (m-samiec) [inline ci/setup-and-build.ps1]: Use $unreal_path or $UNREAL_HOME instead?

File: ci/setup-and-build.ps1 [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    Finish-Event "clone-gdk-plugin" "build-gdk-third-person-shooter-:windows:"
+
+    Start-Event "set-up-gdk-plugin" "build-gdk-third-person-shooter-:windows:"
+        pushd "Game/Plugins/UnrealGDK"
+            # Set the required variables for the GDK's setup script to use
+            $gdk_home = Convert-Path .
+            $msbuild_exe = "${env:ProgramFiles(x86)}\MSBuild\14.0\bin\MSBuild.exe"
+
+            # Invoke the setup script
+            &"$($game_home)\Game\Plugins\UnrealGDK\ci\setup-gdk.ps1"
```
Comment (m-samiec) [inline ci/setup-and-build.ps1]: Use $gdk_home ?

File: ci/setup-and-build.ps1 [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    # Invoke the GDK commandlet to generate schema and snapshot. Note: this needs to be run prior to cooking 
+    Start-Event "generate-schema" "build-gdk-third-person-shooter-:windows:"
+        pushd "UnrealEngine/Engine/Binaries/Win64"
+            Start-Process -Wait -PassThru -NoNewWindow -FilePath ((Convert-Path .) + "\UE4Editor.exe") -ArgumentList @(`
+                "$($game_home)/Game/ThirdPersonShooter.uproject", `
+                "-run=GenerateSchemaAndSnapshots", `
+                "-MapPaths=`"/Maps/TPS-Start_Small`""
+            )
+
+            $core_gdk_schema_path = "$($game_home)\Game\Plugins\UnrealGDK\SpatialGDK\Extras\schema\*"
```
Comment (m-samiec) [inline ci/setup-and-build.ps1]: This logic is very GDK dependent. Perhaps we should invoke some scripts within the gdk directory directly.

File: ci/setup-and-build.ps1 [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                Write-Log "Failed to unzip Unreal Engine. Error: $($zip_proc.ExitCode)"
+                Throw "Failed to unzip Unreal Engine."
+            }
+        popd
+    Finish-Event "download-unreal-engine" "build-gdk-third-person-shooter-:windows:"
+
+    Start-Event "install-unreal-engine-prerequisites" "build-gdk-third-person-shooter-:windows:"
+        # This runs an opaque exe downloaded in the previous step that does *some stuff* that UE needs to occur.
+        # Trapping error codes on this is tricky, because it doesn't always return 0 on success, and frankly, we just don't know what it _will_ return.
+        Start-Process -Wait -PassThru -NoNewWindow -FilePath "$($unreal_path)\Engine\Extras\Redist\en-us\UE4PrereqSetup_x64.exe" -ArgumentList @(`
```
Comment (m-samiec) [inline ci/setup-and-build.ps1]: This uses $unreal_path which is only set below.

File: ci/deploy.ps1 [hunk 6]
```diff
@@ -0,0 +1,39 @@
+. "$PSScriptRoot\common.ps1"
+
+Start-Event "deploy-game" "build-gdk-third-person-shooter-:windows:"
+pushd "spatial"
+    # Get the short commit hash of this build
+    $commit_id = (get-item env:BUILDKITE_COMMIT).Value.Substring(0, 7)
```
Comment (m-samiec) [inline ci/deploy.ps1]: The shooter game project is unlikely to change regularly, perhaps we should use the GDK version here instead?

File: ci/setup-and-build.ps1 (hunk 5 above)
Comment (padhansell) [reply ci/setup-and-build.ps1]: Oh dear lord yes

File: ci/deploy.ps1 (hunk 6 above)
Comment (padhansell) [reply ci/deploy.ps1]: Raised https://improbableio.atlassian.net/browse/UNR-1126 to look in to this

**Pull request merged**